oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -661,22 +660,95 @@ impl Runtime {
                 )
             })?;
 
-        // Try starting the node instance first. If this fails, then directly return the error to
-        // the caller.
-        instance.start()?;
-
-        // If the node was successfully started, insert it in the list of currently running
-        // nodes.
-        self.add_running_node(
-            reference,
-            Node {
-                reference,
-                instance: Some(instance),
+        self.node_start_instance(reference, instance, label, vec![reader])?;
+
+        Ok(())
+    }
+
+    /// Starts a newly created node instance, by first initializing the necessary [`NodeInfo`] data
+    /// structure in [`Runtime`], allowing it to access the provided [`Handle`]s, then calling
+    /// [`Node::start`] on the instance, and finally storing a reference to the running instance
+    /// in [`Runtime::node_instances`] so that it can later be terminated.
+    fn node_start_instance<I>(
+        &self,
+        node_reference: NodeId,
+        mut node_instance: Box<dyn crate::node::Node>,
+        label: &oak_abi::label::Label,
+        initial_handles: I,
+    ) -> Result<(), OakStatus>
+    where
+        I: IntoIterator<Item = Handle>,
+    {
+        // First create the necessary info data structure in the Runtime, otherwise calls that the
+        // node makes to the Runtime may fail.
+        self.add_node_info(
+            node_reference,
+            NodeInfo {
                 label: label.clone(),
-                handles: Mutex::new(vec![reader].into_iter().collect()),
+                handles: Mutex::new(HashSet::new()),
             },
         );
 
-        Ok(())
+        // Make sure that the provided initial handles are tracked in the newly created node from
+        // the start.
+        self.track_handles_in_node(node_reference, initial_handles);
+
+        // Try to start the node instance, and return any errors.
+        let result = node_instance.start();
```

Comment:
The general advice for writing clean code is that concurrency is a separate concern and usually deserves to be handle in a  separate class (a dedicated data structure). I am not sure if this advice is applicable to Rust code, but in any other programming language I've used, I'd write a wrapper class to hold `node_infos` and `node_instances` with high level methods for `get`, `insert`, `remove`, etc. All the concurrency logic (including any necessary syncing between `node_infos` and `node_instances`) would be handled in that class. The only problem I currently see with this approach is that if `node_instance.start()` must absolutely be called **in between** updating `node_infos` and updating `node_instances`. Is that the case?

Reply:
The main point of this PR is so that the node instance can be started so that the relevant data structures (`NodeInfo` in this case) are available to the runtime. I agree with the desire of encapsulating any synchronization in high level accessors, but I'm not entirely sure how to go about it here, perhaps let's discuss separately? If you have any idea in mind and you would like to prototype it in a PR, that would be great!

Reply:
Sure. Let's discuss it separately. What I am thinking of is just a bit of refactoring that can be done separately, after you are done with this PR. And I think it might address some of the comments from @daviddrysdale too (but I am not a true Rustacean yet, so I could be wrong).